src/legacy/server/logging/rotate/log_rotator.js

```diff
@@ -0,0 +1,292 @@
+ * under the License.
+ */
+
+import * as chokidar from 'chokidar';
+import { isMaster } from 'cluster';
+import fs from 'fs';
+import { once, throttle } from 'lodash';
+import { tmpdir } from 'os';
+import { basename, dirname, join, sep } from 'path';
+import { promisify } from 'util';
+
+const mkdirAsync = promisify(fs.mkdir);
+const readdirAsync = promisify(fs.readdir);
+const renameAsync = promisify(fs.rename);
+const statAsync = promisify(fs.stat);
+const unlinkAsync = promisify(fs.unlink);
+const writeFileAsync = promisify(fs.writeFile);
+
+export class LogRotator {
+  constructor(config) {
+    this.logFilePath = config.get('logging.dest');
+    this.interval = 1;
+    this.everyBytes = config.get('logging.rotate.everyBytes');
+    this.keepFiles = config.get('logging.rotate.keepFiles');
+    this.running = false;
+    this.logFileSize = 0;
+    this.isRotating = false;
+    this.throttledRotate = throttle(async () => { await this._rotate(); }, 5000);
+    this.stalker = null;
+    this.usePolling = config.get('logging.rotate.usePolling');
+    this.pollingInterval = config.get('logging.rotate.pollingInterval') * 1000;
+    this.stalkerUsePollingPolicyTestTimeout = null;
+  }
+
+  async start() {
+    if (this.running) {
+      return;
+    }
+
+    // create exit listener for cleanup purposes
+    this._createExitListener();
+
+    // call rotate on startup
+    await this._callRotateOnStartup();
+
+    // init log file size monitor
+    await this._startLogFileSizeMonitor();
+
+    this.running = true;
+  }
+
+  stop() {
+    if (!this.running) {
+      return;
+    }
+
+    // cleanup exit listener
+    this._deleteExitListener();
+
+    // stop log file size monitor
+    this._stopLogFileSizeMonitor();
+
+    this.running = false;
+  }
+
+  async _shouldUsePolling() {
+    try {
+      // Setup a test file in order to try the fs env
+      // and understand if we need to usePolling or not
+      const tempFileDir = tmpdir();
+      const tempFile = join(tempFileDir, 'kbn_log_rotation_use_polling_test_file.log');
+
+      await mkdirAsync(tempFileDir, { recursive: true });
+      await writeFileAsync(tempFile, '');
+
+      const testWatcher = fs.watch(tempFile, { persistent: false });
+
+      return new Promise(async (resolve) => {
+        let fallbackTimeout = null;
+
+        const onResolve = once(async (resolve, completeStatus) => {
+          clearTimeout(fallbackTimeout);
+
+          testWatcher.close();
+          await unlinkAsync(tempFile);
+          resolve(completeStatus);
+        });
+
+        fallbackTimeout = setTimeout(async () => {
+          await onResolve(resolve, true);
+        }, 15000);
+
+        testWatcher.on('change', async () => {
+          await onResolve(resolve, false);
+        });
+
+        testWatcher.on('error', async () => {
+          await onResolve(resolve, true);
+        });
+
+        await writeFileAsync(tempFile, 'test');
+      });
+    } catch {
+      return true;
```

Comment:
It would be great if we could log the reason we reverted to polling here, or the exception if there was one. I could see us introducing a bug that would always result in polling.

Reply:
👍 done